sematic/api/endpoints/runs.py

```diff
@@ -179,6 +198,114 @@ def get_run_endpoint(user: Optional[User], run_id: str) -> flask.Response:
+            HTTPStatus.INTERNAL_SERVER_ERROR,
+        )
+    run.started_at = datetime.datetime.utcnow()
+    save_run(run)
+    payload = dict(
+        content=run.to_json_encodable(),
+    )
+    return flask.jsonify(payload)
+
+
+@sematic_api.route("/api/v1/runs/future_states", methods=["POST"])
+@authenticate
+@retry(_DetectedRunRaceCondition, tries=3, delay=10, jitter=1)
+def update_run_status_endpoint(user: Optional[User]) -> flask.Response:
+    """Update the state of runs based on external job status, and return results"""
+    input_payload: Dict[str, Any] = flask.request.json  # type: ignore
+    if "run_ids" not in input_payload:
+        return jsonify_error(
+            "Call did not contain json with a 'run_ids' key", HTTPStatus.BAD_REQUEST
+        )
+    run_ids = input_payload["run_ids"]
+
+    db_status_dict = get_run_status_details(run_ids)
+    missing_run_ids = set(run_ids).difference(db_status_dict.keys())
+    if len(missing_run_ids) != 0:
+        return jsonify_error(
+            f"Missing runs with ids: {','.join(missing_run_ids)}", HTTPStatus.NOT_FOUND
+        )
+
+    result_list = []
+    for run_id, (future_state, jobs) in db_status_dict.items():
+        new_future_state, message, new_external_jobs = update_run_status(
+            future_state, jobs
+        )
+        run_modified = False
+        run = None
+        if new_external_jobs != jobs:
+            run = get_run(run_id)
+            run.external_jobs = new_external_jobs
+            logger.info("Updating run's external jobs: %s", new_external_jobs)
+            run_modified = True
+        if new_future_state != future_state:
+            run = get_run(run_id) if run is None else run
+            run_modified = True
+            if (
+                run.future_state != future_state.value
+                and run.future_state != new_future_state.value
+            ):
+                # future_state was pulled from the run right before we
+                # started doing our updates. The server logic to this
+                # point hasn't saved any status updates. The run status
+                # must have been changed by a call from the worker or
+                # resolver in the interim.
+                raise _DetectedRunRaceCondition(
+                    f"Run {run_id} was changed from {future_state} to "
+                    f"{run.future_state} out of band of the server. The "
+                    f"server wanted to update the run to {new_future_state}"
+                )
+            run.future_state = new_future_state
+            if message is not None and run.exception is None:
+                run.exception = message
+            logger.info(
+                "Updating run %s from %s to %s. Message: %s",
+                run_id,
+                future_state,
+                new_future_state,
+                message,
+            )
+        if run_modified and run is not None:
```

Comment:
In Rails there is a way to know if a model has been changed since it was loaded from the DB. Unsure if there is something similar in SQLAlchemy.

Reply:
We need to implement something like this. It can't be done without storing some kind of hash/version in the DB and consulting it when you need to know if there was a modification. There are now 3 places that can potentially update a run's state:
(1) the resolver
(2) the worker (to store results, exceptions, etc..)
(3) the server (to store external job updates)

We will eventually add (4) the UI for tags and things

With all these places, you can end up in a situation where entity A reads a run, entity B writes a change, entity A makes a modification to its local copy, and then entity A writes based on its modification, squashing the change from B.

The fix looks something like this:
- readers read a version of the run. They get a hash of the run's state at read-time
- they make a change to the run's state
- they POST to the server two things: here is the new run, and here is the *hash of what I read before modifying*
- the server reads the existing run from the DB, and looks at the hash of what's there. If it is the same as what the writer based its change on, it commits the change. If it differs from what the writer based its change on, the server returns an error indicating that the run has changed.
- if writers get an error that says the run was changed since their last read, they re-read, make the change, and try to post the change again.

This may sound like a lot, but a good bit of it can be hidden behind nice APIs. For the writers, it can look like `self.update_run_field(run_id, exception="foo")`, to update the exception field to `"foo"`, for example. There are some variations on this idea, but you get the gist. Until we do that, we will need to have places here and there where we explicitly re-load the run because it may have been modified elsewhere.

Reply:
Yeah my comment was about detecting in-memory changes, not cross-process.

In Rails you can do something like:
```python
run = get_run(id)
run.touched  # False
run.exception = message
run.touched  # True
save_run(run)
run.touched  # False
```